src/macaron/slsa_analyzer/package_registry/maven_central_registry.py

```diff
@@ -236,3 +239,84 @@ def find_publish_timestamp(self, purl: str) -> datetime:
                 raise InvalidHTTPResponseError(f"The timestamp returned by {url} is invalid") from error
 
         raise InvalidHTTPResponseError(f"Invalid response from Maven central for {url}.")
+
+    @staticmethod
+    def get_artifact_file_name(purl: PackageURL) -> str | None:
+        """Return the artifact file name of the passed PURL based on the Maven registry standard.
+
+        Parameters
+        ----------
+        purl: PackageURL
+            The PURL of the artifact.
+
+        Returns
+        -------
+        str | None
+            The artifact file name, or None if invalid.
+        """
+        if not purl.version:
+            return None
+
+        return purl.name + "-" + purl.version + ".jar"
+
+    def get_artifact_hash(self, purl: PackageURL, hash_algorithm: Any) -> str | None:
+        """Return the hash of the artifact found by the passed purl relevant to the registry's URL.
+
+        Parameters
+        ----------
+        purl: PackageURL
+            The purl of the artifact.
+        hash_algorithm: Any
+            The hash algorithm to use.
+
+        Returns
+        -------
+        str | None
+            The hash of the artifact, or None if not found.
+        """
+        if not purl.namespace:
+            return None
+
+        file_name = MavenCentralRegistry.get_artifact_file_name(purl)
+        if not (purl.version and file_name):
+            return None
+
+        # Maven supports but does not require a sha256 hash of uploaded artifacts.
+        artifact_path = construct_maven_repository_path(purl.namespace, purl.name, purl.version)
+        artifact_url = self.registry_url + "/" + artifact_path + "/" + file_name
+        sha256_url = artifact_url + ".sha256"
+        logger.debug("Search for artifact hash using URL: %s", [sha256_url, artifact_url])
+
+        response = send_get_http_raw(sha256_url, {})
+        sha256_hash = None
+        if response and (sha256_hash := response.text):
+            # As Maven hashes are user provided and not verified they serve as a reference only.
+            logger.debug("Found hash of artifact: %s", sha256_hash)
+
+        try:
+            response = requests.get(artifact_url, stream=True, timeout=40)
+            response.raise_for_status()
+        except requests.exceptions.HTTPError as http_err:
+            logger.debug("HTTP error occurred: %s", http_err)
+            return None
+
+        if response.status_code != 200:
+            return None
+
+        # Download file and compute hash as chunks are received.
```

Comment:
Just a curious question: was there a specific reason why we have to download chunk by chunk at this location?

Reply:
This is the preferred way to download large files. Whether we expect the artefact files to be "large" could be worth discussing.